README.md
# PKCS#11 Reader
*A utility to read PKCS#11 card information from any card and card reader.*

## How to build

The project is based on `pom.xml` files, so you need to have Maven installed and configured. Clone the repo and call:

```shell
    $ maven clean install
```

This will proceed on the main `pom.xml` to clean and build the project.

## How to use

To read information from a card, connect the card reader, insert the card, then call the `ReadCLI` class on the built `jar` with the proper parameters:

```shell
	$ java -jar pkcs11-reader-0.1-SNAPSHOT-jar-with-dependencies.jar ReadCLI
```

The card information will be printed on the standard output, so you can redirect it to any supported output stream.

### Parameters

The following list contains all of the accepted parameters.

* **REQUIRED** `--driver` or `-d`, the library
path for the SSCD (e.g. smart card), must be a valid path
* `--password` or `-p`, the PIN of the card to read information from
* `--encryption` or `-e`, a 16-characters password to encrypt the found information with (see [Encryption](#encryption)).

### Encryption

The read information will be encrypted, if the `--encryption` parameter is specified, with the following settings:

* `AES` Algorithm
* `CBC` Cipher Mode
* `PKCS5Padding` Padding
* the input 16-byte encryption password

### Example

```shell
    $ java -jar pkcs11-reader-0.1-SNAPSHOT-jar-with-dependencies.jar ReadCLI
		--driver="drivers/bit4ipki.dll"
		--password="mypin"
		--encryption="0123456789abcdef"
```

## License

PKCS#11-Reader is released under the LGPL.

    PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
    Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
    Developed by Francesco Pontillo

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Lesser General Public License as published
    by the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see [http://www.gnu.org/licenses/].

### Authors

The original **SD-DSS** project has been commissioned by the European Commission (DG MARKT, Directorate E, Services),
financed under the ISA Work Programme (DG DIGIT) in the framework of the implementation of Services Directive.

**PKCS#11-Reader** has been developed by [Francesco Pontillo](mailto:francescopontillo@gmail.com)
([La Traccia](http://www.latraccia.it/en/)).

### Used libraries

PKCS#11-Reader directly uses the following libraries/modules:

* [**SD-DSS**](https://joinup.ec.europa.eu/software/sd-dss), developed by
[**ARHS Developments S.A.**](http://www.arhs-developments.com) (rue Nicolas Bové 2B, L-1253 Luxembourg), released by
2011 European Commission, Directorate-General Internal Market and Services (DG MARKT), B-1049 Bruxelles/Brussel
under **LGPL v3**.
* [**JCommander**](http://jcommander.org/), developed by [**Cédric Beust**](mailto:cedric@beust.com), released under
**Apache 2.0 license**.

You can find a list of referenced libraries' licenses in the `licenses` directory and the related notices in the
`NOTICE` file in the root of the project.

CHANGELOG.md
# CHANGELOG
-----------

## 0.0.1-SNAPSHOT (2014-03-18)
* First release

src/test/java/it/latraccia/pkcs11/reader/test/Const.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.test;

/**
 * Date: 18/03/14
 * Time: 16.59
 *
 * @author Francesco Pontillo
 */
public class Const {
    public static final String ENCRYPTION_PASSWORD = "0123456789abcdef";

    public static class Actalis {
        public static final String DECRYPTED = "SERIAL_NUMBER=65235973081121999\n" +
                "VALIDITY_FROM=2013-09-25T13:32:50.000+02:00\n" +
                "VALIDITY_TO=2018-09-25T13:32:50.000+02:00\n" +
                "ISSUER_CN=Actalis CA per certificati CNS G1\n" +
                "ISSUER_OU=CNS Certification Service Provider\n" +
                "ISSUER_O=Actalis S.p.A./03358520967\n" +
                "ISSUER_C=IT\n" +
                "SUBJECT_CN=\"NTZRCC81H16L418X/2444024000002384.cifJCU3a1W7XVru2h79d2sR3gjE=\"\n" +
                "SUBJECT_OU=Ulss 8 Asolo\n" +
                "SUBJECT_O=Ulss 8 Asolo\n" +
                "SUBJECT_C=IT\n\n";

        public static final String ENCRYPTED = "r7sFAjPWcivgKjSHO2iU6I2W8Ze6OUvGM05Dry8Ew9VEJq+M4gr3vcbFD3NYAnz" +
                "YOOUcoVPLDeeovXrY5EpqthBYRKKTttm4QadlEQyzjidHXDttBzoSptDDGaf/wi+F0v2LCxxw1Su2nAn3HojcfgKNu4BjMnPB5Zl5Qt" +
                "ruMC9Z12qe7J5ByM9r+q8g0R/FJNO+C8DAPXi6pMVP3WSK9J2QNtSpwgOwFqvTtRUI9XEYIDGSGLfbaWdE56zLKU3biJsZlLRw5sVyh" +
                "eDvwCJvNP11unpppVCvaWrjVxa5J8HkQmfFHjcoaGg4P1nVRgtr9w05SlMHWsWDsAfvScqUrjUJrbMcgzuviaUyWitMjG7oI3adcTK4" +
                "ajAFF7aah9Vf+ZnDV7CufW1e9MRinJ+wWlsMjuSi7ljTH7MeugubaFU0XbcsroFdD5bRDtVRT0hUpi/A138hf3AIDdxx2zwoemR7ymb" +
                "Kq8sgEv9BHyJuZEFJXCUlhVdkozrKOzakqKCBdDGSX7krwtspDWcP6HDBbQ==";
    }

    public static class InfoCert {
        public static final String DECRYPTED = "SERIAL_NUMBER=3094481\n" +
                "VALIDITY_FROM=2013-01-30T12:35:32.000+01:00\n" +
                "VALIDITY_TO=2016-01-30T01:00:00.000+01:00\n" +
                "ISSUER_CN=InfoCert Firma Qualificata\n" +
                "ISSUER_OU=Certificatore Accreditato\n" +
                "ISSUER_SERIALNUMBER=07945211006\n" +
                "ISSUER_O=INFOCERT SPA\n" +
                "ISSUER_C=IT\n" +
                "SUBJECT_CN=Rocco Antezza\n" +
                "SUBJECT_DNQ=20101498441\n" +
                "SUBJECT_SERIALNUMBER=IT:NTZRCC81H16L418X\n" +
                "SUBJECT_GIVENNAME=ROCCO\n" +
                "SUBJECT_SURNAME=ANTEZZA\n" +
                "SUBJECT_O=U.L.SS. 8 ASOLO/00896810264\n" +
                "SUBJECT_C=IT\n\n";

        public static final String ENCRYPTED = "Q8WI/90Rzo54Z7vXxIBz6sBW+sqbf+owWOv7Cz2L0lm9XRV9k0BAMcCIHKQCLcBfkgSWBSlMr" +
                "86wSyQ4wvWm4ASNCnQFllz3yEP3bAGnn11YdG1f1i6JYcjuk1RauVqim3c8kBt6jeCdofb5IKTjA6b4kOVXLlUBmRjIsETgF1W3QpO1i" +
                "yIV6YXf7pBIb0iYp7RB3zfMmGo681m6b6WKExj6ZmYJxghEfga8EgMfjFDjLOo/3ZCcsELZ3rPugOkdLWR8At6DlFLsGIL682f/pEq3K" +
                "DMgodUOjus3qgKv7o4KU7bVtddHAeZT+wq5d5SOAF4NGwWLh861qAtqz3sYhk5k7icVSDL1OeirqgQvojwuvKGhmjY8KW166JsBCpxc9" +
                "BeBxOqqjBi5IsDFXo4+OmSs8PTrNpEVMwdOE5UPcpUe54U3dcafDxNiU+zdEduRb7roZSJX7Kv1ffXGxDZbrshKI7jMjfvCuzw4nozSt" +
                "y2aairyJJ3hQIBDuI4PaoYzz53KD7ngTuA6rla0ZhhoMho6V+tyd2en8ZW06TZDUtZQ5XCirTeCmWN4tgwSAprfrvu27HL2l8vFc8x2d" +
                "o0TaQ==";
    }
}


src/test/java/it/latraccia/pkcs11/reader/test/util/AESTest.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.test.util;

import it.latraccia.pkcs11.reader.util.AESUtil;
import org.junit.Test;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

import static org.junit.Assert.*;

/**
 * Generic AESTest class for testing the encryption and decryption of several string combos.
 * Simply extend this class and call the super constructor with the desired values as parameters.
 *
 * @author Francesco Pontillo
 */
public abstract class AESTest {
    String encryptionPassword;
    String expectedEncrypted;
    String expectedDecrypted;

    protected AESTest(String encrypted, String decrypted, String encryptionPassword) {
        this.encryptionPassword = encryptionPassword;
        this.expectedEncrypted = encrypted;
        this.expectedDecrypted = decrypted;
    }

    @Test
    public void encrypt() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException {
        String encrypted = AESUtil.encryptString(expectedDecrypted, encryptionPassword);
        assertEquals(this.expectedEncrypted, encrypted);
    }

    @Test
    public void decrypt() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException {
        String decrypted = AESUtil.decryptString(expectedEncrypted, encryptionPassword);
        assertEquals(this.expectedDecrypted.trim(), decrypted.trim());
    }

    @Test(expected = IllegalArgumentException.class)
    public void longerPassword() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException {
        AESUtil.decryptString(expectedEncrypted, encryptionPassword + "a-salt");
    }

    @Test(expected = IllegalArgumentException.class)
    public void shorterPassword() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException {
        AESUtil.decryptString(expectedEncrypted, "short-password");
    }

    @Test
    public void wrongPassword() throws NoSuchPaddingException, InvalidKeyException, NoSuchAlgorithmException, IllegalBlockSizeException, BadPaddingException, InvalidAlgorithmParameterException {
        String decrypted = null;
        try {
            decrypted = AESUtil.decryptString(expectedEncrypted, "a-wrong-password");
        } catch (javax.crypto.BadPaddingException exception) {
            // ignore, it's ok if we end up here because of a bad password
        }
        assertNotSame(expectedDecrypted, decrypted);
    }
}


src/test/java/it/latraccia/pkcs11/reader/test/util/ActalisAESTest.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.test.util;

import it.latraccia.pkcs11.reader.test.Const;
import org.junit.experimental.categories.Category;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/**
 * Test class for Actalis string encryption/decryption.
 *
 * @author Francesco Pontillo
 */
@RunWith(JUnit4.class)
@Category(value = AESTest.class)
public class ActalisAESTest extends AESTest {
    public ActalisAESTest() {
        super(Const.Actalis.ENCRYPTED, Const.Actalis.DECRYPTED, Const.ENCRYPTION_PASSWORD);
    }
}


src/main/java/it/latraccia/pkcs11/reader/cli/ReadCLI.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.cli;

import eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry;
import it.latraccia.pkcs11.reader.cli.argument.ReadingArgs;
import com.beust.jcommander.JCommander;
import eu.europa.ec.markt.dss.signature.token.Pkcs11SignatureToken;
import it.latraccia.pkcs11.reader.util.AESUtil;
import it.latraccia.pkcs11.reader.util.DSSPrivateKeyUtil;

import javax.crypto.BadPaddingException;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.util.List;

/**
 * Command Line Interface for reading and encrypting information from a PKCS#11 card.
 *
 * @author Francesco Pontillo
 */
public class ReadCLI {

    public static void main(String[] args) {
        ReadingArgs readingArgs = new ReadingArgs();
        String returnThis = null;

        try {
            new JCommander(readingArgs, args);

            // Get the keys
            List<DSSPrivateKeyEntry> keys = getKeys(readingArgs);
            // Get a string representation
            String stringKeys = DSSPrivateKeyUtil.DSSKeysToString(keys);
            // Possibly encrypt the resulting string
            returnThis = cryptMeMaybe(stringKeys, readingArgs);
        } catch (Exception e) {
            returnThis = "ERROR\n" + e.getMessage();
        }

        // Print the resulting string
        System.out.println(returnThis);
    }

    /**
     * Get the {@link eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry}s from the PKCS#11 card.
     * @param readingArgs The {@link it.latraccia.pkcs11.reader.cli.argument.ReadingArgs} to get the card password from.
     * @return A list of found keys.
     * @throws KeyStoreException If the card could not be read.
     */
    private static List<DSSPrivateKeyEntry> getKeys(ReadingArgs readingArgs) throws KeyStoreException {
        Pkcs11SignatureToken token = new Pkcs11SignatureToken(
                readingArgs.getDriver().getAbsolutePath(), readingArgs.getPassword().toCharArray());
        List<DSSPrivateKeyEntry> keys = token.getKeys();
        return keys;
    }

    /**
     * Encrypts the input {@link java.lang.String} (or not) according to the presence of the encryption password
     * in the {@link it.latraccia.pkcs11.reader.cli.argument.ReadingArgs}.
     *
     * @param cryptMe The {@link java.lang.String} that will possibly be encrypted.
     * @param readingArgs The {@link it.latraccia.pkcs11.reader.cli.argument.ReadingArgs} to get the encryption
     *                    password from.
     * @return The {@link java.lang.String} to be printed.
     *
     * @throws NoSuchPaddingException
     * @throws InvalidKeyException
     * @throws NoSuchAlgorithmException
     * @throws IllegalBlockSizeException
     * @throws BadPaddingException
     * @throws InvalidAlgorithmParameterException
     */
    private static String cryptMeMaybe(String cryptMe, ReadingArgs readingArgs) throws NoSuchPaddingException,
                                                                                       InvalidKeyException,
                                                                                       NoSuchAlgorithmException,
                                                                                       IllegalBlockSizeException,
                                                                                       BadPaddingException,
                                                                                       InvalidAlgorithmParameterException {
        if (readingArgs.getEncryptionPassword() != null) {
            return AESUtil.encryptString(cryptMe, readingArgs.getEncryptionPassword());
        }
        return cryptMe;
    }
}


src/main/java/it/latraccia/pkcs11/reader/util/AESUtil.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.util;

import org.apache.commons.codec.binary.Base64;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;

/**
 * Date: 18/03/14
 * Time: 13.07
 *
 * @author Francesco Pontillo
 */
public class AESUtil {
    public static String encryptString(String clearText, String password) throws NoSuchPaddingException,
                                                                                 NoSuchAlgorithmException,
                                                                                 InvalidAlgorithmParameterException,
                                                                                 InvalidKeyException,
                                                                                 BadPaddingException,
                                                                                 IllegalBlockSizeException {
        String encrypted;
        byte[] key = password.getBytes();
        // Check the length of the password
        if (key.length != 16) {
            throw new IllegalArgumentException("Invalid password length.");
        }
        byte[] value = clearText.getBytes();

        // Encrypt with AES/CBC/PKCS5Padding
        SecretKeySpec skeySpec = new SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, skeySpec, new IvParameterSpec(new byte[16]));
        byte[] encryptedBytes = cipher.doFinal(value);

        // Encode the bytes in base64
        encrypted = Base64.encodeBase64String(encryptedBytes);
        return encrypted;
    }

    public static String decryptString(String encryptedText, String password) throws NoSuchPaddingException,
                                                                                     NoSuchAlgorithmException,
                                                                                     InvalidAlgorithmParameterException,
                                                                                     InvalidKeyException,
                                                                                     BadPaddingException,
                                                                                     IllegalBlockSizeException {
        String decrypted = null;
        byte[] key = password.getBytes();
        if (key.length != 16) {
            throw new IllegalArgumentException("Invalid key size.");
        }
        byte[] value = Base64.decodeBase64(encryptedText);

        // Decrypt with AES/CBC/PKCS5Padding
        SecretKeySpec skeySpec = new SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.DECRYPT_MODE, skeySpec, new IvParameterSpec(new byte[16]));
        byte[] original = cipher.doFinal(value);
        decrypted = new String(original);

        return decrypted;
    }
}


src/main/java/it/latraccia/pkcs11/reader/util/FileUtil.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.util;


import java.io.*;
import java.net.URL;

public class FileUtil {

    /**
     * Search a file in the resources, and return its absolute path if it exists,
     * otherwise return a null {@link String}.
     *
     * @param fileName The name of the file in the resources
     * @return The path {@link String} of the file, if any, or null
     */
    public static String getFileInResources(String fileName) {
        URL fileURL = FileUtil.class.getClassLoader().getResource(fileName);
        String filePath = null;
        if (fileURL != null) {
            filePath = fileURL.getFile();
        }
        return filePath;
    }

    /**
     * Search for a file in an absolute path, and always returns a {@link String},
     * even if the file does not exist.
     *
     * @param fileName The file absolute path
     * @return The path {@link String} of the file
     */
    public static String getFileInAbsolutePath(String fileName) {
        File file = new File(fileName);
        return file.getAbsolutePath();
    }

    /**
     * Search for a given file path:
     *  - handling the fileName as an absolute path
     *  - fallback in resources, returns it if it exists
     *
     * @param fileName File name or path of a file in resources or an absolute path
     * @return  The file absolute path as a {@link String}, if it was found anywhere
     *          in the absolute path or in the resources
     */
    public static String getFileInAbsolutePathOrResources(String fileName) throws FileNotFoundException {
        File file = new File(fileName);
        String filePath = file.getAbsolutePath();
        if (!file.exists()) {
             filePath = getFileInResources(fileName);
        }
        return filePath;
    }
}


src/main/java/it/latraccia/pkcs11/reader/util/StringUtil.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.util;

/**
 * Date: 18/03/14
 * Time: 12.32
 *
 * @author Francesco Pontillo
 */
public class StringUtil {

    /**
     * Check if a {@link String} is null or empty.
     * @param string The {@link String} to check
     * @return true if the {@link String} is null or empty, false otherwise
     */
    public static boolean isNullOrEmpty(String string) {
        return (string == null || string.isEmpty());
    }
}


src/main/java/it/latraccia/pkcs11/reader/cli/argument/ReadingArgs.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.cli.argument;

import com.beust.jcommander.Parameter;
import com.beust.jcommander.Parameters;
import com.beust.jcommander.converters.FileConverter;

import java.io.File;

/**
 * Mapping class containing all the definitions for the {@link it.latraccia.pkcs11.reader.cli.ReadCLI} arguments.
 * Every argument hereby contained will be automaticall mapepd by {@link com.beust.jcommander.JCommander}.
 *
 * @author Francesco Pontillo
 */
@Parameters(separators = "=")
public class ReadingArgs {

    @Parameter(
            names = {"-d", "--driver"},
            required = true,
            converter = it.latraccia.pkcs11.reader.cli.argument.converter.FileConverter.class,
            description = "The PKCS#11 card driver path.")
    private File driver;

    @Parameter(
            names = {"-p", "--password"},
            description = "Password for the PKCS#11 card.")
    private String password;

    @Parameter(
            names = {"-e", "--encrypt"},
            description = "Password for the data encryption.")
    private String encryptionPassword;

    public File getDriver() {
        return driver;
    }

    public void setDriver(File driver) {
        this.driver = driver;
    }

    public String getPassword() {
        return password;
    }

    public void setPassword(String password) {
        this.password = password;
    }

    public String getEncryptionPassword() {
        return encryptionPassword;
    }

    public void setEncryptionPassword(String encryptionPassword) {
        this.encryptionPassword = encryptionPassword;
    }
}


src/main/java/it/latraccia/pkcs11/reader/cli/argument/converter/FileConverter.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.cli.argument.converter;

import com.beust.jcommander.IStringConverter;
import it.latraccia.pkcs11.reader.util.FileUtil;

import java.io.File;
import java.io.FileNotFoundException;

/**
 * Converter for arguments representing a {@link java.io.File} path that may be located in a "resources" folder
 * or in an absolute path.
 *
 * @author Francesco Pontillo
 */
public class FileConverter implements IStringConverter<File> {
    @Override
    public File convert(String value) {
        try {
            return new File(FileUtil.getFileInAbsolutePathOrResources(value));
        } catch (FileNotFoundException e) {
            e.printStackTrace();
        }
        return null;
    }
}


src/main/java/it/latraccia/pkcs11/reader/util/DSSPrivateKeyUtil.java
/*
 * PKCS#11-Reader, a utility to read PKCS#11 card information from any card and card reader.
 * Copyright (C) 2014 La Traccia http://www.latraccia.it/en/
 * Developed by Francesco Pontillo
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see [http://www.gnu.org/licenses/].
 */

package it.latraccia.pkcs11.reader.util;

import eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry;
import org.joda.time.DateTime;
import org.joda.time.format.DateTimePrinter;

import java.security.cert.X509Certificate;
import java.util.Calendar;
import java.util.GregorianCalendar;
import java.util.List;

/**
 * Date: 18/03/14
 * Time: 12.32
 *
 * @author Francesco Pontillo
 */
public class DSSPrivateKeyUtil {

    /**
     * Get the subject DN {@link String} from a certificate.
     * @param cert A {@link java.security.cert.X509Certificate} to read the subject DN from
     * @return The {@link String} representing the subject for the {@link java.security.cert.X509Certificate}
     */
    public static String getSubjectDN(X509Certificate cert) {
        String subjectDN = cert.getSubjectDN().getName();
        int dnStartIndex = subjectDN.indexOf("CN=") + 3;
        if (dnStartIndex > 0 && subjectDN.indexOf(",", dnStartIndex) > 0) {
            subjectDN = subjectDN.substring(dnStartIndex, subjectDN.indexOf(",", dnStartIndex)) + " (SN:"
                    + cert.getSerialNumber() + ")";
        }
        return subjectDN;
    }

    /**
     * Create a {@link java.lang.String} representation for a {@link eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry}.
     *
     * @param key the input {@link eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry}.
     * @return A {@link java.lang.String} representation.
     */
    public static StringBuilder DSSKeyToString(DSSPrivateKeyEntry key) {
        StringBuilder keyBuilder = new StringBuilder();
        X509Certificate certificate = key.getCertificate();
        // The serial number
        keyBuilder.append("SERIAL_NUMBER=");
        keyBuilder.append(certificate.getSerialNumber().toString());
        keyBuilder.append("\n");
        // The validity dates
        keyBuilder.append("VALIDITY_FROM=");
        keyBuilder.append(new DateTime(certificate.getNotBefore()).toString());
        keyBuilder.append("\n");
        keyBuilder.append("VALIDITY_TO=");
        keyBuilder.append(new DateTime(certificate.getNotAfter()).toString());
        keyBuilder.append("\n");

        // Get the issuer data and prepend them with "ISSUER_"
        String prependIssuer = "ISSUER_";
        String issuer = certificate.getIssuerDN().getName();
        String[] issuerData = issuer.split(", ");
        for (String data : issuerData) {
            keyBuilder.append(prependIssuer + data);
            keyBuilder.append("\n");
        }

        // Get the subject data and prepend them with "SUBJECT_"
        String prependSubject = "SUBJECT_";
        String subject = certificate.getSubjectDN().getName();
        String[] subjectData = subject.split(", ");
        for (String data : subjectData) {
            keyBuilder.append(prependSubject + data);
            keyBuilder.append("\n");
        }

        return keyBuilder;
    }

    /**
     * Returns the {@link java.lang.String} representation of a {@link java.util.List} of {@link eu.europa.ec.markt.dss.signature.token.DSSPrivateKeyEntry}s.
     * @param keys The list of keys that will be serialized.
     * @return The resulting serialization.
     */
    public static String DSSKeysToString(List<DSSPrivateKeyEntry> keys) {
        StringBuilder serializedStringBuilder = new StringBuilder();
        for (DSSPrivateKeyEntry key : keys) {
            serializedStringBuilder.append(DSSPrivateKeyUtil.DSSKeyToString(key));
            serializedStringBuilder.append("\n");
        }
        return serializedStringBuilder.toString();
    }
}
